File: app/src/main/java/com/nmc/android/ui/ScanActivity.java
package com.nmc.android.ui;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;

import com.nextcloud.client.preferences.AppPreferences;
import com.nmc.android.interfaces.OnDocScanListener;
import com.nmc.android.interfaces.OnFragmentChangeListener;
import com.owncloud.android.R;
import com.owncloud.android.databinding.ActivityScanBinding;
import com.owncloud.android.datamodel.OCFile;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.operations.CreateFolderIfNotExistOperation;
import com.owncloud.android.ui.activity.FileActivity;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import io.scanbot.sdk.ScanbotSDK;

public class ScanActivity extends FileActivity implements OnFragmentChangeListener, OnDocScanListener {

    protected static final String FRAGMENT_SCAN_TAG = "SCAN_FRAGMENT_TAG";
    protected static final String FRAGMENT_EDIT_SCAN_TAG = "EDIT_SCAN_FRAGMENT_TAG";
    protected static final String FRAGMENT_CROP_SCAN_TAG = "CROP_SCAN_FRAGMENT_TAG";
    protected static final String FRAGMENT_SAVE_SCAN_TAG = "SAVE_SCAN_FRAGMENT_TAG";

    //default path to upload the scanned document
    //if user doesn't select any location then this will be the default location
    public static final String DEFAULT_UPLOAD_SCAN_PATH = OCFile.ROOT_PATH + "Scans" + OCFile.PATH_SEPARATOR;

    protected static final String TAG = "ScanActivity";
    private static final String EXTRA_REMOTE_PATH = "com.nmc.android.ui.scan_activity.extras.remote_path";

    private ActivityScanBinding binding;
    private ScanbotSDK scanbotSDK;

    public static final List<Bitmap> originalScannedImages = new ArrayList<>();//list with original bitmaps
    public static final List<Bitmap> filteredImages = new ArrayList<>();//list with bitmaps applied filters
    public static final List<Integer> scannedImagesFilterIndex = new ArrayList<>();//list to maintain the state of
    // applied filter index when device rotated

    @Inject AppPreferences appPreferences;

    private String remotePath;
    //flag to avoid checking folder existence whenever user goes to save fragment
    //we will make it true when the operation finishes first time
    private boolean isFolderCheckOperationFinished;

    public static void openScanActivity(Context context, String remotePath, int requestCode) {
        Intent intent = new Intent(context, ScanActivity.class);
        intent.putExtra(EXTRA_REMOTE_PATH, remotePath);
        ((AppCompatActivity) context).startActivityForResult(intent, requestCode);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Inflate and set the layout view
        binding = ActivityScanBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        remotePath = getIntent().getStringExtra(EXTRA_REMOTE_PATH);
        originalScannedImages.clear();
        filteredImages.clear();
        scannedImagesFilterIndex.clear();
        initScanbotSDK();
        setupToolbar();
        setupActionBar();
    }

    private void setupActionBar() {
        ActionBar actionBar = getDelegate().getSupportActionBar();

        if (actionBar != null) {
            //required for NMC
            actionBar.setBackgroundDrawable(new ColorDrawable(getResources().getColor(R.color.bg_default, null)));
            actionBar.setDisplayHomeAsUpEnabled(true);
            //NMC Customization
            viewThemeUtils.files.themeActionBar(this, actionBar, false);
        }
    }

    public String getRemotePath() {
        return remotePath;
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        createScanFragment(savedInstanceState);
    }

    private void createScanFragment(Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            ScanDocumentFragment scanDocumentFragment = ScanDocumentFragment.newInstance(TAG);
            onReplaceFragment(scanDocumentFragment, FRAGMENT_SCAN_TAG, false);
        } else {
            getSupportFragmentManager().findFragmentByTag(FRAGMENT_SCAN_TAG);
        }
    }

    @Override
    public void onReplaceFragment(Fragment fragment, String tag, boolean addToBackStack) {
        //only during replacing save scan fragment
        if (tag.equalsIgnoreCase(FRAGMENT_SAVE_SCAN_TAG)) {
            checkAndCreateFolderIfRequired();
        }

        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.scan_frame_container, fragment, tag);
        if (addToBackStack) {
            transaction.addToBackStack(tag);
        }
        transaction.commit();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressHandle();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        onBackPressHandle();
    }

    private void onBackPressHandle() {
        Fragment editScanFragment = getSupportFragmentManager().findFragmentByTag(FRAGMENT_EDIT_SCAN_TAG);
        Fragment cropScanFragment = getSupportFragmentManager().findFragmentByTag(FRAGMENT_CROP_SCAN_TAG);
        Fragment saveScanFragment = getSupportFragmentManager().findFragmentByTag(FRAGMENT_SAVE_SCAN_TAG);
        if (cropScanFragment != null || saveScanFragment != null) {
            int index = 0;
            if (cropScanFragment instanceof CropScannedDocumentFragment) {
                index = ((CropScannedDocumentFragment) cropScanFragment).getScannedDocIndex();
            }
            onReplaceFragment(EditScannedDocumentFragment.newInstance(index), FRAGMENT_EDIT_SCAN_TAG, false);
        } else if (editScanFragment != null) {
            createScanFragment(null);
        } else {
            super.onBackPressed();
        }
    }

    private void initScanbotSDK() {
        scanbotSDK = new ScanbotSDK(this);
    }

    public ScanbotSDK getScanbotSDK() {
        return scanbotSDK;
    }

    @Override
    public void addScannedDoc(Bitmap file) {
        if (file != null) {
            originalScannedImages.add(file);
            filteredImages.add(file);
            scannedImagesFilterIndex.add(0);//no filter by default
        }
    }

    @Override
    public List<Bitmap> getScannedDocs() {
        return filteredImages;
    }

    @Override
    public boolean removedScannedDoc(Bitmap file, int index) {
        //removed the filter applied index also when scanned document is removed
        if (scannedImagesFilterIndex.size() > 0 && scannedImagesFilterIndex.size() > index) {
            scannedImagesFilterIndex.remove(index);
        }
        if (originalScannedImages.size() > 0 && file != null) {
            originalScannedImages.remove(index);
        }
        if (filteredImages.size() > 0 && file != null) {
            filteredImages.remove(index);
            return true;
        }
        return false;
    }

    @Override
    public Bitmap replaceScannedDoc(int index, Bitmap newFile, boolean isFilterApplied) {
        //only update the original bitmap if no filter is applied
        if (!isFilterApplied && originalScannedImages.size() > 0 && newFile != null && index >= 0 && originalScannedImages.size() - 1 >= index) {
            originalScannedImages.set(index, newFile);
        }
        if (filteredImages.size() > 0 && newFile != null && index >= 0 && filteredImages.size() - 1 >= index) {
            return filteredImages.set(index, newFile);
        }
        return null;
    }

    @Override
    public void replaceFilterIndex(int index, int filterIndex) {
        if (scannedImagesFilterIndex.size() > 0 && scannedImagesFilterIndex.size() > index) {
            scannedImagesFilterIndex.set(index, filterIndex);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        List<Fragment> fragmentList = getSupportFragmentManager().getFragments();
        for (Fragment fragment : fragmentList) {
            fragment.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    protected void checkAndCreateFolderIfRequired() {
        //if user is coming from sub-folder then we should not check for existence as folder will be available
        if (!TextUtils.isEmpty(remotePath) && !remotePath.equals(OCFile.ROOT_PATH)) {
            return;
        }

        //no need to do any operation if its already finished earlier
        if (isFolderCheckOperationFinished) {
            return;
        }

        String lastRemotePath = appPreferences.getUploadScansLastPath();

        //create the default scan folder if it doesn't exist or if user has not selected any other folder
        if (lastRemotePath.equalsIgnoreCase(ScanActivity.DEFAULT_UPLOAD_SCAN_PATH)) {
            getFileOperationsHelper().createFolderIfNotExist(lastRemotePath, false);
            return;
        }

        //if last saved remote path is not root path then we have to check if the folder exist or not
        if (!lastRemotePath.equals(OCFile.ROOT_PATH)) {
            getFileOperationsHelper().createFolderIfNotExist(lastRemotePath, true);
        }
    }

    @Override
    public void onRemoteOperationFinish(RemoteOperation operation, RemoteOperationResult result) {
        super.onRemoteOperationFinish(operation, result);
        if (operation instanceof CreateFolderIfNotExistOperation) {
            //we are only handling callback when we are checking if folder exist or not to update the UI
            //in case the folder doesn't exist (user has deleted)
            if (!result.isSuccess() && result.getCode() == RemoteOperationResult.ResultCode.FILE_NOT_FOUND) {
                Fragment saveScanFragment = getSupportFragmentManager().findFragmentByTag(FRAGMENT_SAVE_SCAN_TAG);
                if (saveScanFragment != null && saveScanFragment.isVisible()) {
                    //update the root path in preferences as well
                    //so that next time folder issue won't come
                    appPreferences.setUploadScansLastPath(OCFile.ROOT_PATH);
                    //if folder doesn't exist then we have to set the remote path as root i.e. fallback mechanism
                    ((SaveScannedDocumentFragment) saveScanFragment).setRemoteFilePath(OCFile.ROOT_PATH);
                }
            }
            isFolderCheckOperationFinished = true;
        }
    }
}

File: app/src/main/java/com/nmc/android/scans/ScanActivity.kt
package com.nmc.android.scans

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.nextcloud.utils.extensions.getParcelableArgument
import com.nmc.android.interfaces.OnDocScanListener
import com.nmc.android.interfaces.OnFragmentChangeListener
import com.nmc.android.scans.ScanDocumentFragment.Companion.newInstance
import com.owncloud.android.R
import com.owncloud.android.databinding.ActivityScanBinding
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.resources.status.OCCapability
import com.owncloud.android.operations.CreateFolderIfNotExistOperation
import com.owncloud.android.ui.activity.FileActivity
import com.owncloud.android.ui.dialog.setupEncryption.SetupEncryptionDialogFragment
import com.owncloud.android.ui.helpers.FileOperationsHelper
import com.owncloud.android.utils.DisplayUtils
import io.scanbot.sdk.ScanbotSDK
import androidx.core.graphics.drawable.toDrawable

class ScanActivity : FileActivity(), OnFragmentChangeListener, OnDocScanListener {
    private lateinit var binding: ActivityScanBinding
    lateinit var scanbotSDK: ScanbotSDK

    var remoteFile: OCFile? = null
        private set

    // flag to avoid checking folder existence whenever user goes to save fragment
    // we will make it true when the operation finishes first time
    private var isFolderCheckOperationFinished = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Inflate and set the layout view
        binding = ActivityScanBinding.inflate(layoutInflater)
        setContentView(binding.root)
        remoteFile = intent.getParcelableArgument(EXTRA_REMOTE_PATH, OCFile::class.java)
        originalScannedImages.clear()
        filteredImages.clear()
        scannedImagesFilterIndex.clear()
        initScanbotSDK()
        setupToolbar()
        setupActionBar()
        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)
    }

    private fun setupActionBar() {
        val actionBar = delegate.supportActionBar
        actionBar?.let {
            it.setBackgroundDrawable(resources.getColor(R.color.bg_default, null).toDrawable())
            it.setDisplayHomeAsUpEnabled(true)
            viewThemeUtils.files.themeActionBar(this, it, false)
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        createScanFragment(savedInstanceState)
    }

    private fun createScanFragment(savedInstanceState: Bundle?) {
        if (savedInstanceState == null) {
            val scanDocumentFragment = newInstance(TAG)
            onReplaceFragment(scanDocumentFragment, FRAGMENT_SCAN_TAG, false)
        } else {
            supportFragmentManager.findFragmentByTag(FRAGMENT_SCAN_TAG)
        }
    }

    override fun onReplaceFragment(fragment: Fragment, tag: String, addToBackStack: Boolean) {
        // only during replacing save scan fragment
        if (tag.equals(FRAGMENT_SAVE_SCAN_TAG, ignoreCase = true)) {
            checkAndCreateFolderIfRequired()
        }

        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.scan_frame_container, fragment, tag)
        if (addToBackStack) {
            transaction.addToBackStack(tag)
        }
        transaction.commit()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressHandle()
        }
        return super.onOptionsItemSelected(item)
    }

    private val onBackPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            onBackPressHandle()
        }
    }

    private fun onBackPressHandle() {
        val editScanFragment = supportFragmentManager.findFragmentByTag(FRAGMENT_EDIT_SCAN_TAG)
        val cropScanFragment = supportFragmentManager.findFragmentByTag(FRAGMENT_CROP_SCAN_TAG)
        val saveScanFragment = supportFragmentManager.findFragmentByTag(FRAGMENT_SAVE_SCAN_TAG)
        if (cropScanFragment != null || saveScanFragment != null) {
            var index = 0
            if (cropScanFragment is CropScannedDocumentFragment) {
                index = cropScanFragment.getScannedDocIndex()
            }
            onReplaceFragment(EditScannedDocumentFragment.newInstance(index), FRAGMENT_EDIT_SCAN_TAG, false)
        } else if (editScanFragment != null) {
            createScanFragment(null)
        } else {
            finish()
        }
    }

    private fun initScanbotSDK() {
        scanbotSDK = ScanbotSDK(this)
    }

    override fun addScannedDoc(file: Bitmap?) {
        file?.let {
            originalScannedImages.add(it)
            filteredImages.add(it)
            scannedImagesFilterIndex.add(0) // no filter by default
        }
    }

    override fun getScannedDocs(): List<Bitmap> {
        return filteredImages
    }

    override fun removedScannedDoc(file: Bitmap?, index: Int): Boolean {
        //removed the filter applied index also when scanned document is removed
        if (scannedImagesFilterIndex.isNotEmpty() && scannedImagesFilterIndex.size > index) {
            scannedImagesFilterIndex.removeAt(index)
        }
        if (originalScannedImages.isNotEmpty() && file != null) {
            originalScannedImages.removeAt(index)
        }
        if (filteredImages.isNotEmpty() && file != null) {
            filteredImages.removeAt(index)
            return true
        }
        return false
    }

    override fun replaceScannedDoc(index: Int, newFile: Bitmap?, isFilterApplied: Boolean): Bitmap? {
        //only update the original bitmap if no filter is applied
        if (!isFilterApplied && originalScannedImages.isNotEmpty() && newFile != null && index >= 0 && originalScannedImages.size - 1 >= index) {
            originalScannedImages[index] = newFile
        }
        if (filteredImages.isNotEmpty() && newFile != null && index >= 0 && filteredImages.size - 1 >= index) {
            return filteredImages.set(index, newFile)
        }
        return null
    }

    override fun replaceFilterIndex(index: Int, filterIndex: Int) {
        if (scannedImagesFilterIndex.isNotEmpty() && scannedImagesFilterIndex.size > index) {
            scannedImagesFilterIndex[index] = filterIndex
        }
    }

    private fun checkAndCreateFolderIfRequired() {
        val remotePath = remoteFile?.remotePath

        //if user is coming from sub-folder then we should not check for existence as folder will be available
        if (!TextUtils.isEmpty(remotePath) && remotePath != OCFile.ROOT_PATH) {
            return
        }

        //no need to do any operation if its already finished earlier
        if (isFolderCheckOperationFinished) {
            return
        }

        val lastRemotePath = appPreferences.uploadScansLastPath

        //create the default scan folder if it doesn't exist or if user has not selected any other folder
        if (lastRemotePath.equals(DEFAULT_UPLOAD_SCAN_PATH, ignoreCase = true)) {
            fileOperationsHelper.createFolderIfNotExist(lastRemotePath, false)
            return
        }

        //if last saved remote path is not root path then we have to check if the folder exist or not
        if (lastRemotePath != OCFile.ROOT_PATH) {
            fileOperationsHelper.createFolderIfNotExist(lastRemotePath, true)
        }
    }

    // NMC-3670
    // check if selected folder is encrypted and e2ee is configured or not
    fun checkEncryption(file: OCFile, resultListener: (success: Boolean) -> Unit) {
        // get file from storage to have the encrypted information
        // as we are making OCFile without the flag {see-> SaveScannedDocumentFragment.setRemoteFilePath()}
        var remoteFile = storageManager.getFileByEncryptedRemotePath(file.remotePath)
        // there can be case where the remoteFile can be null
        if (remoteFile == null) {
            remoteFile = file
        }

        if (!remoteFile.isEncrypted) {
            resultListener(true)
            return
        }

        if (remoteFile.isEncrypted) {
            val user = user.orElseThrow { RuntimeException() }

            // check if e2e app is enabled
            val ocCapability: OCCapability = storageManager
                .getCapability(user.accountName)

            if (ocCapability.endToEndEncryption.isFalse ||
                ocCapability.endToEndEncryption.isUnknown
            ) {
                DisplayUtils.showSnackMessage(this, R.string.end_to_end_encryption_not_enabled)
                resultListener(false)
                return
            }
            // check if keys are stored
            if (FileOperationsHelper.isEndToEndEncryptionSetup(this, user)) {
                resultListener(true)
            } else {
                val setupEncryptionDialogFragment = SetupEncryptionDialogFragment.newInstance(user, -1)
                supportFragmentManager.setFragmentResultListener(
                    SetupEncryptionDialogFragment.RESULT_REQUEST_KEY,
                    this
                ) { requestKey, result ->
                    if (requestKey == SetupEncryptionDialogFragment.RESULT_REQUEST_KEY) {
                        resultListener(
                            !result.getBoolean(SetupEncryptionDialogFragment.RESULT_KEY_CANCELLED, false)
                                && result.getBoolean(SetupEncryptionDialogFragment.SUCCESS, false)
                        )
                    }
                }
                setupEncryptionDialogFragment.show(
                    supportFragmentManager,
                    SetupEncryptionDialogFragment.SETUP_ENCRYPTION_DIALOG_TAG
                )
            }
        }
    }

    override fun onRemoteOperationFinish(operation: RemoteOperation<*>?, result: RemoteOperationResult<*>) {
        super.onRemoteOperationFinish(operation, result)
        if (operation is CreateFolderIfNotExistOperation) {
            //we are only handling callback when we are checking if folder exist or not to update the UI
            //in case the folder doesn't exist (user has deleted)
            if (!result.isSuccess && result.code == RemoteOperationResult.ResultCode.FILE_NOT_FOUND) {
                val saveScanFragment = supportFragmentManager.findFragmentByTag(FRAGMENT_SAVE_SCAN_TAG)
                if (saveScanFragment != null && saveScanFragment.isVisible) {
                    //update the root path in preferences as well
                    //so that next time folder issue won't come
                    appPreferences.uploadScansLastPath = OCFile.ROOT_PATH
                    //if folder doesn't exist then we have to set the remote path as root i.e. fallback mechanism
                    (saveScanFragment as SaveScannedDocumentFragment).setRemoteFilePath(OCFile.ROOT_PATH)
                }
            }
            // NMC-4746 fix
            else if (result.isSuccess) {
                val saveScanFragment = supportFragmentManager.findFragmentByTag(FRAGMENT_SAVE_SCAN_TAG)
                if (saveScanFragment != null && saveScanFragment.isVisible) {
                    // when folder creation is success update the path in save fragment
                    (saveScanFragment as SaveScannedDocumentFragment).setRemoteFilePath(appPreferences.uploadScansLastPath)
                }
            }
            isFolderCheckOperationFinished = true
        }
    }

    companion object {
        const val FRAGMENT_SCAN_TAG: String = "SCAN_FRAGMENT_TAG"
        const val FRAGMENT_EDIT_SCAN_TAG: String = "EDIT_SCAN_FRAGMENT_TAG"
        const val FRAGMENT_CROP_SCAN_TAG: String = "CROP_SCAN_FRAGMENT_TAG"
        const val FRAGMENT_SAVE_SCAN_TAG: String = "SAVE_SCAN_FRAGMENT_TAG"

        // default path to upload the scanned document
        // if user doesn't select any location then this will be the default location
        const val DEFAULT_UPLOAD_SCAN_PATH: String = OCFile.ROOT_PATH + "Scans" + OCFile.PATH_SEPARATOR

        const val TAG: String = "ScanActivity"
        private const val EXTRA_REMOTE_PATH = "com.nmc.android.scans.scan_activity.extras.remote_path"

        @JvmField
        val originalScannedImages: MutableList<Bitmap> = ArrayList() //list with original bitmaps

        @JvmField
        val filteredImages: MutableList<Bitmap> = ArrayList() //list with bitmaps applied filters

        @JvmField
        val scannedImagesFilterIndex: MutableList<Int> = ArrayList() //list to maintain the state of
        // applied filter index when device rotated

        @JvmStatic
        fun openScanActivity(context: Context, remoteFile: OCFile, requestCode: Int) {
            val intent = Intent(context, ScanActivity::class.java)
            intent.putExtra(EXTRA_REMOTE_PATH, remoteFile)
            (context as AppCompatActivity).startActivityForResult(intent, requestCode)
        }
    }
}
